tsdb/engine/tsm1/engine.go

```diff
@@ -228,7 +258,8 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) writeSnapshot() error {
+// WriteSnapshot will snapshot the cache and write a new TSM file with its contents, releasing the snapshot when done.
+func (e *DevEngine) WriteSnapshot(async bool) error {
```

Comment:
Compaction runs in a single goroutine. In our testing with TSM1 we found that we became CPU bound by a single CPU (the one doing the compaction). Having this means that two snapshots can run at the same time on different CPUs.

@otoolep based on your comment we already have a problem with the `compact` goroutine not sending any information back to everything else.

In fact, this brings up a good point, which is that in many places we're logging errors and continuing on despite the fact that the errors should probably be irrecoverable (i.e. panic).